refactor(utxo): refactor utxo output script creation

dimxy opened the pull request:
Closes #1270 
As suggested in the issue, script_type field is added to Address structure so we do not need the extra script type param in output_script() any more. Now the script type is determined right at parsing input address strings. For that we always need coin prefixes when we parse addresses as strings, so the standard Address::from_str() was replaced on Address::from_legacyaddress() function, with coin prefixes as params. 
The output_script() fn is now universally used everywhere when output script should be created from an address

There are still few cases when we need the standard from_str fn (w/o prefixes) to parse addresses as strings, f.e. the convertaddress rpc. For this a new LegacyAddress type is added which supports the from_str() method.

artemii235 commented on mm2src/coins/lightning/ln_events.rs:
```diff
@@ -219,7 +220,7 @@ fn sign_funding_transaction(
         .map_err(|e| SignFundingTransactionError::Internal(e.to_string()))?;
 
-    let prev_script = Builder::build_p2pkh(&my_address.hash);
+    let prev_script = output_script(my_address);
```
I suppose this change will break sighash calculation, as `output_script` will return `build_witness_script` for Segwit address. As per https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#native-p2wpkh, the scriptCode is still `OP_DUP OP_HASH160 hash160(pubkey) OP_EQUALVERIFY OP_CHECKSIG` for Native P2WPKH input signing.

There are also test failures related to similar changes in other places: https://github.com/KomodoPlatform/komodo-defi-framework/actions/runs/6096406569/job/16541897281?pr=1960#step:5:1529

dimxy replied:
Ah, sorry, I fixed that segwit signing code for this case in my original PR but did not make it here.
(I ran tests but apparently there are more tests in nested crates that I missed)

artemii235 commented on mm2src/coins/qrc20.rs:
```diff
@@ -536,7 +537,7 @@ impl Qrc20Coin {
         let key_pair = self.utxo.priv_key_policy.activated_key_or_err()?;
 
-        let prev_script = ScriptBuilder::build_p2pkh(&my_address.hash);
+        let prev_script = output_script(my_address);
```
Same here, it seems to break signing if Segwit is used.

artemii235 commented on mm2src/coins/utxo.rs:
```diff
@@ -1826,7 +1826,7 @@ where
     };
 
-    let prev_script = Builder::build_p2pkh(&my_address.hash);
+    let prev_script = output_script(my_address);
```
Same here, it seems to break signing if Segwit is used.

artemii235 commented on mm2src/coins/utxo/qtum_delegation.rs:
```diff
@@ -290,7 +295,7 @@ impl QtumCoin {
             })?;
 
-        let prev_script = ScriptBuilder::build_p2pkh(&my_address.hash);
+        let prev_script = output_script(my_address);
```
Same here, it seems to break signing if Segwit is used.

artemii235 commented on mm2src/coins/utxo/slp.rs:
```diff
@@ -1661,7 +1663,7 @@ impl MmCoin for SlpToken {
             })?;
 
-            let prev_script = ScriptBuilder::build_p2pkh(&my_address.hash);
+            let prev_script = output_script(my_address);
```
Same here, it seems to break signing if Segwit is used (it's not supported for BCH/SLP, though).

dimxy replied:
I guess segwit address will never be created based on coin data?

artemii235 replied:
Yes, it won't be created unless BCH supports Segwit one day 🙂

artemii235 commented on mm2src/coins/utxo/utxo_tests.rs:
```diff
@@ -3381,7 +3395,7 @@ fn test_split_qtum() {
         _ => coin.as_ref().conf.signature_version,
     };
-    let prev_script = Builder::build_p2pkh(&p2pkh_address.hash);
+    let prev_script = output_script(p2pkh_address);
```
Same here, it seems to break signing if Segwit is used.

artemii235 reviewed (changes requested):
Thanks for the PR! I have few changes requests, all related to signing.

dimxy commented:
Added missed fix for segwit signing code (thanks @artemii235 for noticing that).
The explanation is: as now a unified output_script fn is used for all cases when prev_script is created, for segwit we pass scriptPubKey (not scriptCode as before) to the signing code. The segwit signing code creates scriptCode itself. I believe this is more similar to how the original signing code in daemons does this.

(Seems tests are better now. There are still few tests failing but I believe because of this PR)

shamardy commented:
> (Seems tests are better now. There are still few tests failing but I believe because of this PR)

Yeah, failing tests now are not due to changes in this PR.

artemii235 commented on mm2src/mm2_bitcoin/rpc/src/v1/types/mod.rs:
```diff
@@ -20,8 +20,9 @@ pub use self::hash::{H160, H256, H264};
                             SignedTransactionOutput, Transaction, TransactionInput, TransactionInputEnum,
-                            TransactionInputScript, TransactionOutput, TransactionOutputScript,
-                            TransactionOutputWithAddress, TransactionOutputWithScriptData, TransactionOutputs};
+                            TransactionInputScript, /*TransactionOutput,*/ TransactionOutputScript,
+                            TransactionOutputWithAddress,
+                            TransactionOutputWithScriptData /*, TransactionOutputs*/};
```
If `TransactionOutput` and `TransactionOutputs` are deleted, it is preferable to remove them from imports instead of commenting out.

artemii235 commented on mm2src/coins/qrc20.rs:
```diff
@@ -1491,12 +1492,16 @@ pub struct Qrc20FeeDetails {
-        .map_to_mm(WithdrawError::InvalidAddress)?;
+    let to_addr = UtxoAddress::from_legacyaddress(
+        &req.to,
+        coin.as_ref().conf.pub_addr_prefix,
+        coin.as_ref().conf.pub_t_addr_prefix,
+        coin.as_ref().conf.p2sh_addr_prefix,
+        coin.as_ref().conf.p2sh_t_addr_prefix,
```
What do you think about creating the following struct and using it instead to make code a bit shorted/avoid duplication?
```rust
struct AddressPrefixes {
    pub_addr_prefix: u8,
    pub_t_addr_prefix: u8,
    p2sh_addr_prefix: u8,
    p2sh_t_addr_prefix: u8,
}
```
So this and other `from_legacyaddress` calls will look like
```rust
UtxoAddress::from_legacyaddress(&req.to, coin.as_ref().conf.address_prefixes)
```

dimxy replied:
yes, I also thought about a struct like that or array. I will do this, thank you

dimxy replied:
What would you say about using Vec<u8> for prefixes, like it is in daemons codebases. So such a vec would look like [t_addr_prefix, p2pkh_prefix] with t_addr_prefix optional.
```
pub struct Address {
    /// The prefixes of the p2pkh address of size 1 or 2
    pub p2pkh_prefix: Vec<u8>,
    /// The prefixes of the p2sh address of size 1 or 2
    pub p2sh_prefix: Vec<u8>,
    ...
}
```
Functions would accept slices in their params.
I think it would be easier to deal with such types (less variables or members)
Also we do not have t_addr_prefix in most cases, so we do not process it with this approach.
Plus this is consistent with other prefixes representations like xpub

dimxy replied:
Added NetworkAddressPrefixes and AddressPrefixes structs to encapsulate prefixes

artemii235 commented on mm2src/coins/qrc20/qrc20_tests.rs:
```diff
@@ -150,7 +152,7 @@ fn test_validate_maker_payment() {
     assert_eq!(
         *coin.utxo.derivation_method.unwrap_single_addr(),
-        "qUX9FGHubczidVjWPCUWuwCUJWpkAtGCgf".into()
+        Address::from_legacyaddress("qUX9FGHubczidVjWPCUWuwCUJWpkAtGCgf", 120, 0, 50, 0).unwrap()
```
I think you can use `coin.as_ref().conf.pub_addr_prefix/...` here and other similar lines. Also, QTUM regtest P2SH prefix is 110.

dimxy replied:
I tried as_ref but I guess deref here is okay, the alternative is to use '&Address::from_legacyaddress(..).
Also instead of using consts I changed it to coin.as_ref().conf.address_prefixes

artemii235 commented on mm2src/mm2_bitcoin/keys/src/legacyaddress.rs:
```diff
@@ -0,0 +1,117 @@
+    }
+}
+
+impl From<&'static str> for LegacyAddress {
+    fn from(s: &'static str) -> Self { s.parse().unwrap() } // TODO: dangerous unwrap?
+}
```
You can likely add `#cfg[test]` to make it test-only and avoid occasional usage and panic in production code.

dimxy replied:
I could not use cfg[test] as this code used in tests in other modules but I added Default instead

artemii235 commented on mm2src/coins/utxo/utxo_common_tests.rs:
```diff
@@ -195,25 +200,28 @@ pub(super) fn get_morty_hd_transactions_ordered(tx_hashes: &[&str]) -> Vec<Trans
-        "RYPz6Lr4muj4gcFzpMdv3ks1NCGn3mkDPN".into(),
-        "RJeDDtDRtKUoL8BCKdH7TNCHqUKr7kQRsi".into(),
-        "RQHn9VPHBqNjYwyKfJbZCiaxVrWPKGQjeF".into(),
+        Address::from_legacyaddress("RG278CfeNPFtNztFZQir8cgdWexVhViYVy", 60, 0, 85, 0).unwrap(),
+        Address::from_legacyaddress("RYPz6Lr4muj4gcFzpMdv3ks1NCGn3mkDPN", 60, 0, 85, 0).unwrap(),
+        Address::from_legacyaddress("RJeDDtDRtKUoL8BCKdH7TNCHqUKr7kQRsi", 60, 0, 85, 0).unwrap(),
+        Address::from_legacyaddress("RQHn9VPHBqNjYwyKfJbZCiaxVrWPKGQjeF", 60, 0, 85, 0).unwrap(),
```
It is preferred to use constants like `const KMD_P2PKH: u8 = 60;` and so on.

dimxy replied:
added contants

artemii235 reviewed (commented):
Few more suggestions, please consider them as non-blocking.

dimxy commented:
I hope, I have fixed the requests above

laruh commented:
@dimxy Hi, this PR started to have conflicts.

dimxy commented:
rebased on the latest dev

dimxy commented:
merged recent dev changes

shamardy commented on mm2src/coins/utxo/utxo_builder/utxo_coin_builder.rs:
```diff
@@ -230,15 +230,15 @@ where
         checksum_type: conf.checksum_type,
         hrp: conf.bech32_hrp.clone(),
         addr_format,
+        script_type: AddressScriptType::P2PKH,
```
This can be `AddressScriptType::P2WPKH` depending on the address format. Without fixing this, the change I suggested in the comment above will not work.

dimxy replied:
I think even better would be to use builder:
```
    let my_address = AddressBuilder {
        prefixes: conf.address_prefixes.p2pkh.clone(),
        hash: AddressHashEnum::AddressHash(key_pair.public().address_hash()),
        checksum_type: conf.checksum_type,
        hrp: conf.bech32_hrp.clone(),
        addr_format,
        is_pay2pk: AddressScriptType::P2PKH,
    }.build();```

shamardy replied:
I agree, Builder pattern is a good fit for this case, we should have `AddressBuilder::new()` function to initiate the building process, with setter functions for the address parameters.

dimxy replied:
added AddressBuilder in [87c438d](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1960/commits/87c438d11d2f42244943b846d6b1fa29b358d9a4) with extra creation data checks (which allowed to find an issue in swaps with segwit [a38ed54](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1960/commits/a38ed54a911a2f9c2afe8ee6d1078c6a71126bb8): Segwit is not supported in swaps so addr_format must be converted from segwit to legacy here)

shamardy commented on mm2src/coins/utxo/utxo_builder/utxo_conf_builder.rs:
```diff
@@ -51,10 +53,27 @@ impl<'a> UtxoConfBuilder<'a> {
+        if pub_t_addr_prefix != 0 {
+            p2pkh_prefixes.push(pub_t_addr_prefix);
+        }
+        p2pkh_prefixes.push(pub_addr_prefix);
```
```rust
drop_mutability!(p2pkh_prefixes);
```

dimxy replied:
fixed

shamardy commented on mm2src/coins/utxo/utxo_builder/utxo_conf_builder.rs:
```diff
@@ -51,10 +53,27 @@ impl<'a> UtxoConfBuilder<'a> {
+        if p2sh_t_addr_prefix != 0 {
+            p2sh_prefixes.push(p2sh_t_addr_prefix);
+        }
+        p2sh_prefixes.push(p2sh_addr_prefix);
```
`drop_mutability` here too

dimxy replied:
fixed

shamardy commented on mm2src/mm2_bitcoin/script/src/sign.rs:
```diff
@@ -737,7 +746,7 @@ mod tests {
             shielded_outputs: vec![],
             zcash: false,
-            posv: true,
+            posv: true, // TODO: Is this intended? In coins file for coin "tBLK", with this prefix 111, the field "pos" is true, not "posv".
```
Yes, this is intended as it was part of this PR https://github.com/KomodoPlatform/komodo-defi-framework/pull/1815 for Reddcoin and Potcoin.

shamardy reviewed (commented):
Thank you for the fixes! Next iteration, will do one final review after this one!

dimxy commented:
fixed review notes, added AddressBuilder and divided the witness script creation fn into two functions separately for wpkh and wsh (for better address hash checking)

shamardy commented on mm2src/coins/lightning/ln_events.rs:
```diff
@@ -532,7 +532,13 @@ impl LightningEventHandler {
+            let change_destination_script = match Builder::build_p2wpkh(my_address.hash()) {
+                Ok(script) => script.to_bytes().take().into(),
+                Err(_) => {
+                    error!("Error building change script");
```
Please try to make errors as informative as possible, so it's better to include the error and the address in the message.

dimxy replied:
fixed in [64dbf87](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1960/commits/64dbf873bbcc5a92d86e3bc107a9ad7d6db3b715)

shamardy commented on mm2src/coins/qrc20.rs:
```diff
@@ -1537,12 +1536,10 @@ pub struct Qrc20FeeDetails {
-    let is_p2pkh = to_addr.prefix == conf.pub_addr_prefix && to_addr.t_addr_prefix == conf.pub_t_addr_prefix;
-    if !is_p2pkh {
+    if !to_addr.is_p2pkh() {
         let error = "QRC20 can be sent to P2PKH addresses only".to_owned();
```
Maybe we should change this message also to be `p2pkh` and `p2wpkh`

dimxy replied:
It seems that qrc20 cannot go to segwit addresses and this is what this check is supposed to do.
Maybe my is_p2pkh() is confusing and better to return true only for p2pkh (not p2pkh or p2wpkh)

shamardy commented on mm2src/mm2_bitcoin/keys/src/address/address_builder.rs:
```diff
@@ -0,0 +1,110 @@
+    /// Builds Address with the script type corresponding to p2sh or p2wsh
+    pub fn build_as_sh(&self) -> Result<Address, String> { self.build(AddressBuilderOption::BuildScriptHash) }
+
+    pub fn build(&self, build_option: AddressBuilderOption) -> Result<Address, String> {
```
To follow the builder pattern, `build_option: AddressBuilderOption` should be part of `AddressBuilder` struct and you should add either a function `pub fn with_options(mut self, build_option: AddressBuilderOption)` or 2 functions instead, `fn as_pkh(mut self)` and `as_sh(mut self)` that take a mut self too. We should also have `new` function and other setter functions for other fields.

ref. https://rust-unofficial.github.io/patterns/patterns/creational/builder.html

dimxy replied:
fixed in [64dbf87](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1960/commits/64dbf873bbcc5a92d86e3bc107a9ad7d6db3b715)

shamardy reviewed (commented):
Thanks you for the fixes! Next review iteration!

shamardy commented:
@mariocynicys you can give this PR a look/review since changes here will affect your PR https://github.com/KomodoPlatform/komodo-defi-framework/pull/2053 a lot and this is to be merged first.

shamardy commented on mm2src/mm2_bitcoin/keys/src/address/address_builder.rs:
```diff
@@ -5,70 +5,110 @@ use {Address, AddressFormat, AddressHashEnum, AddressPrefixes, AddressScriptType
+        }
+    }
+
+    /*pub fn with_hrp(&mut self, hrp: String) -> Self {
+        self.hrp = Some(hrp);
+        self
+    }
+
+    pub fn with_address_prefixes(&mut self, prefixes: NetworkAddressPrefixes) -> Self {
+        self.prefixes = prefixes;
+        self
+    }*/
```
I guess you forgot to remove this commented out block.

dimxy replied:
done

shamardy commented on mm2src/mm2_bitcoin/script/src/builder.rs:
```diff
@@ -39,12 +39,26 @@ impl Builder {
+                .push_opcode(Opcode::OP_0)
+                .push_bytes(wpkh_hash.as_ref())
+                .into_script()),
+            _ => Err(Error::WitnessHashMismatched),
```
Please match all enum variants to remember handling any added variants in the future.

dimxy replied:
done

shamardy commented on mm2src/mm2_bitcoin/script/src/builder.rs:
```diff
@@ -39,12 +39,26 @@ impl Builder {
+                .push_opcode(Opcode::OP_0)
+                .push_bytes(wsh_hash.as_ref())
+                .into_script()),
+            _ => Err(Error::WitnessHashMismatched),
```
Same here.

dimxy replied:
done

shamardy commented on mm2src/coins/utxo.rs:
```diff
@@ -1903,15 +1901,13 @@ where
+    match address.script_type() {
+        AddressScriptType::P2PKH => Builder::build_p2pkh(address.hash()),
+        AddressScriptType::P2SH => Builder::build_p2sh(address.hash()),
+        AddressScriptType::P2WPKH => Builder::build_p2wpkh(address.hash()).expect("valid p2wpkh"),
+        AddressScriptType::P2WSH => Builder::build_p2wsh(address.hash()).expect("valid p2wsh"),
```
A missed dev error here will lead to panicking which is not good but better than building the wrong output script and sending to a wrong address. I would say the best solution is always to propagate the errors.

dimxy replied:
Added error propagation in [354c970](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1960/commits/354c970e3409d0e31629e40337b5d9819891adcd) - this led to not a few changes in the callers though

shamardy commented on mm2src/coins/utxo/utxo_builder/utxo_coin_builder.rs:
```diff
@@ -229,16 +229,18 @@ where
+    )
+    .as_pkh()
+    .build()
+    .expect("valid address props");
```
Same here and some other places too.

shamardy commented on mm2src/coins/utxo/utxo_withdraw.rs:
```diff
@@ -102,16 +103,20 @@ where
+                Ok(script) => Ok(script),
+                Err(e) => MmError::err(WithdrawError::InternalError(e.to_string())),
+            },
+            _ => Ok(Builder::build_p2pkh(self.sender_address().hash())),
```
Please cover all variants explicitly, for instance if we were to add taproot address in the future the compiler won't ask us to fix this and will build it as `p2pkh` if the dev forgets to check it.

shamardy reviewed (commented):
Next review iteration! Probably the last review iteration unless there are some minor things found during the last code check :)

mariocynicys reviewed (commented):
Thanks!
I only have a couple of nits here.

dimxy commented on mm2src/mm2_bitcoin/keys/src/address_prefixes.rs:
```diff
@@ -0,0 +1,184 @@
+#[derive(Debug, Clone, Eq, Hash, PartialEq, Default)]
+pub struct AddressPrefixes {
+    data: Vec<u8>,
+}
```
Renamed into AddressPrefix and other vars 'prefixes' to 'prefix' in [d7f6cdb](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1960/commits/d7f6cdb56fb62a2996bfb9354ad6354a88647c1e)
initially I used plural as it is an array but I agree singular is better

dimxy commented on mm2src/mm2_bitcoin/keys/src/address_prefixes.rs:
```diff
@@ -0,0 +1,184 @@
+pub struct ConstPrefixes<'a> {
+    p2pkh: &'a [u8],
+    p2sh: &'a [u8],
+}
```
Thank you, got rid from ConstPrefixes
(I did not want to use lazy_static at the first place, but there's no reason not to use it)

dimxy commented on mm2src/mm2_bitcoin/keys/src/address.rs:
```diff
@@ -330,154 +262,232 @@ impl Address {
+                cash_address.encode().expect("A valid address").fmt(f)
+            },
+            AddressFormat::Standard => LegacyAddress::new(&self.hash, self.prefixes.clone(), self.checksum_type)
+                .to_string()
```
fixed in [3434e2b](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1960/commits/3434e2ba72a022118becdc20604d28747303d018)

dimxy commented on mm2src/mm2_bitcoin/keys/src/address.rs:
```diff
@@ -330,154 +262,232 @@ impl Address {
+        match &self.addr_format {
+            AddressFormat::Segwit => {
+                SegwitAddress::new(&self.hash, self.hrp.clone().expect("Segwit address should have an hrp"))
+                    .to_string()
```
thank you, fixed in [3434e2b](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1960/commits/3434e2ba72a022118becdc20604d28747303d018)

shamardy reviewed (approved):
🔥

mariocynicys reviewed (approved):
Thanks, LGTM!

shamardy commented:
This is just a refactor PR and it's needed for other PRs, so sec-review can be done in release PR.
